### PR TITLE
Add background processing with traits futures

### LINKED ISSUE #10: Possible topic to address: background processing
The [traits-futures](https://github.com/enthought/traits-futures) package makes this pretty simple, and it is a common thing that scientific GUIs face.  Maybe having something like a menu item which runs face detection across a directory in the background and notifies when done (or better a progress bar somewhere).

### PR DESCRIPTION
**Pull request by siddhantwahal**
Based on feature/add_step_5.2

Closes #10.
Contributes to #29.

Adds background processing with traits futures. Opening early to get feedback. 

After:
traits_futures

### CODE REVIEW

File: stage5.2_fuller_application/pycasa/ui/image_folder_editor.py [hunk 1]
```diff
@@ -34,3 +35,9 @@ def _get_name(self):
 
     def _get_tooltip(self):
         return self.obj.path
+
+    @observe("closing")
```
Comment (siddhantwahal) [inline stage5.2_fuller_application/pycasa/ui/image_folder_editor.py:39]: This is necessary. If the editor is closed while a computation is active (executing or cancelling), we get the following traceback:
```
Traceback (most recent call last):
  File "/Users/swahal/.edm/envs/ets_tutorial/lib/python3.6/site-packages/traits/trait_notifiers.py", line 342, in __call__
    self.handler(*args)
  File "/Users/swahal/.edm/envs/ets_tutorial/lib/python3.6/site-packages/traitsui/qt4/editor.py", line 86, in _e
nabled_changed
    self._enabled_changed_helper(self.control, enabled)
  File "/Users/swahal/.edm/envs/ets_tutorial/lib/python3.6/site-packages/traitsui/qt4/editor.py", line 95, in _e
nabled_changed_helper
    control.setEnabled(enabled)
RuntimeError: Internal C++ object (PySide2.QtWidgets.QPushButton) already deleted.
```

Note that `future.cancel()` doesn't immediately stop the future. It only moves it to a cancelling state.

**Comment by jonathanrocher** [comment]
animation looks amazing but code looks non-trivial. One thing I would like to ask is to move this to a stage 6 folder.

**Comment by siddhantwahal** [comment]
> Animation looks amazing but code looks non-trivial.
Agreed. Attempted to simplify by:
- no longer vetoing if computation is active
- removing the "Cancel" button
- using a `CallFuture` instead of an `IterationFuture`, which means no more generators and `yield`s

> One thing I would like to ask is to move this to a stage 6 folder.

This is built on top of 5.2, so waiting for that to mature before copying things into Stage 6.

File: stage5.2_fuller_application/pycasa/ui/image_folder_editor.py (hunk 1 above)
Comment (siddhantwahal) [reply stage5.2_fuller_application/pycasa/ui/image_folder_editor.py]: Not relevant anymore.

**Comment by siddhantwahal** [comment]
@siddhantwahal: Note that this is Stage 7.

File: stage5.2_fuller_application/pycasa/model/image_folder.py [hunk 2]
```diff
@@ -24,6 +29,12 @@ class ImageFolder(HasStrictTraits):
 
     data_updated = Event
 
+    traits_executor = Instance(TraitsExecutor)
+
+    future = Instance(CallFuture)
+
+    executor_idle = Property(Bool, observe="future.done")
```
Comment (jonathanrocher) [inline stage5.2_fuller_application/pycasa/model/image_folder.py:36]: Is that the new way of doing `depends_on`?

**Review by jonathanrocher — APPROVED** [review]
Looks neat! Would be cool to have time to have you speak to it!

**Pull request closed without merging**